### PR TITLE
allow configuration of one single password per circle

### PR DESCRIPTION
**Pull request by ArtificialOwl**
- [x] Managing circle's settings from the command line
- [x] Allow multiple configuration to use this feature (in order):
  * Enforcing password protection if the Enforce password protection is enabled.
  * Enforcing password protection if the Circle App is configured to: 
       `./occ config:app:set --value '1' circles enforce_password`
  * Enforcing password protection if the Circle's settings is configured to:
      `./occ circles:manage:setting y8JczmViR3zZraCXhbpfR8v4x8s5A8c enforce_password true`

- [x] Allow the use of a single password when sharing a new file to external members of a circle,
- [x] Allow the use of a single password when adding a new external member to a circle,
- [x] confirm behavior in globalscale env.
- [x] providing API and Feature Request for Contacts
- [x] filtering some setting before sending ocs data about circle

### display settings from a `circleId`

```
$ ./occ circles:manage:setting y8JczmViR3zZraCXhbpfR8v4x8s5A8c 
{
    "enforce_password": "true",
    "password_single_enabled": "true",
    "password_single": "3|$argon2id$v=19$m=65536,t=4,p=1$NktudFp4UmluSGZtRGhEeA$VJuD1vhKuKG+oFFT1nlmXHU3O3aTUw1ZFkjW8Vw9gxU"
}
```

### Add/Edit setting from a `circleId`

```
$ ./occ circles:manage:setting y8JczmViR3zZraCXhbpfR8v4x8s5A8c enforce_password true
$ ./occ circles:manage:setting y8JczmViR3zZraCXhbpfR8v4x8s5A8c password_single_enable true
$ ./occ circles:manage:setting y8JczmViR3zZraCXhbpfR8v4x8s5A8c password_single my_clear_password
```

_Note: when setting the param `password_single` the clear password will be instantly be hashed_

### Remote a setting from a `circleId`

```
$ ./occ circles:manage:setting y8JczmViR3zZraCXhbpfR8v4x8s5A8c enforce_password --unset
```

### CODE REVIEW

**Review by mejo- — APPROVED** [review]

**Review by mejo- — COMMENTED** [review]
Sorry, accidently approved despite having comments/questions. Changed to "comment" now.

**Review by mejo- — COMMENTED** [review]
again

File: lib/Model/Circle.php [hunk 1]
```diff
@@ -872,6 +892,26 @@ public function importFromDatabase(array $data, string $prefix = ''): INC22Query
 
 		$this->getManager()->manageImportFromDatabase($this, $data, $prefix);
 
+
+		// TODO: deprecated in NC27, remove those (17) lines that was needed to finalise migration to 24
+		// if password is not hashed (pre-22), hash it and update new settings in DB
+		$curr = $this->get('password_single', $this->getSettings());
+		if (strlen($curr) > 1 && strlen($curr) < 64) {
```
[the remark this answers is not in the document]
Comment (ArtificialOwl) [reply lib/Model/Circle.php:899]: true, who will use a 1-char password ? :)

File: lib/Service/CircleService.php [hunk 2]
```diff
@@ -290,21 +294,34 @@ public function updateConfig(string $circleId, int $config): array {
 	 * @throws RequestBuilderException
 	 * @throws UnknownRemoteException
 	 */
-	public function updateName(string $circleId, string $name): array {
+	public function updateSetting(string $circleId, string $setting, ?string $value): array {
```
[the remark this answers is not in the document]
Comment (ArtificialOwl) [reply lib/Service/CircleService.php:297]: I am fine with it. 99% of the people will use the front-end to set those settings (with pre-defined keys)

**Pull request merged**